### PR TITLE
Feat/sm/check token

### LINKED ISSUE #40: Create a refresh route
In the event that the AT that is on the frontend has expired, the backend should return a 403 when the FE requests for a resource. On returning the 403 the frontend needs to hit this route by sending the old AT and the RT.

Call the route `/refresh` 
Return a 200 along with the AT and RT

### LINKED ISSUE #41: Create a check route
When the user logs into the app, the app needs to decided whether it needs to render the AuthenticatedApp or the UnAuthenticatedApp. The way it would do this is in the `useEffect` of the app, it will make a request to the `/check` route.

Below are the scenarios,

1. If the AT isn't present then render Unaunthenticated app.
2. If the AT is present, then send it to the `/check` route. If it returns a 403, call `/refresh`. If theres no RT -> Unauth App. If RT isnt valid then -> unauth app.
3. Return a 200 and update the new AT/RT.

After which the authenticated app will be rendered, which will route to the respective room or home. This handles the case when the user refreshes on the page.

What do you think?

### PR DESCRIPTION
**Pull request by s-munro**
# Check and Refresh endpoints
## What:
- `/api/auth/check` endpoint created
- `/api/auth/refresh` endpoint created
- check validates `accessToken` 
- Two new middlewares made

Regarding the middlewares: originally I just did all of the checking logic in it for the `/check` endpoint.  The endpoint was literally just an empty res.send afterwards.  Refactored to move the actual "check" logic into the endpoint, with some basic type checking in the middleware.  Lmk if you think this is right - I was having a hard time drawing the line between what should be middleware and what should be in the controller.

Regarding the refresh middleware, I did the same thing.  "refresh" logic only requires basically one function call though, so most of the additional logic for this one is int the middleware, with only the "refresh" itself in the endpoint.  Once again lmk.  It's 3am I'm a sleepy boy

### CODE REVIEW

File: vb-be/src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -38,16 +38,15 @@ export class AuthService {
...
   }
 
-  async createTokens(user: UserType): Promise<string[]> {
+  async createTokens(user: ITokenUser): Promise<string[]> {
     if (!user) {
       throw new ErrorHandler(HttpStatus.Error, ErrorText.TokenError);
     }
 
     try {
-      const { email, id } = user;
```
Comment (s-munro) [inline vb-be/src/modules/auth/auth.service.ts]: `id` was coming in as `undefined` every time, due to how we originally create the tokens on login.  Removed `id` here for now, but will fix everything tomorrow afternoon.  The culprit:
```
const userObject: UserType = {
    displayName: user.data.display_name,
    email: user.data.email,
    href: user.data.href,
    uri: user.data.uri,
    image: user.data.images.length > 0 ? user.data.images[0].url : null,
    username: user.data.id,
};

const vbUser = await this.userService.findOne(userObject.email);

if (!vbUser) await this.userService.create(userObject);

const [accessToken, refreshToken] = await this.authService.createTokens(userObject);
```

There's no id on the userObject, so user.id was coming in as `undefined`.  This should instead be vbUser - I'm not sure if `create` returns the object as well though, or if it just creates the object and returns undefined. We just need to make sure we pass in a variable that has an ID and all of the info, regardless of if its through finding or creating a user.  Any thoughts?

File: vb-be/src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -85,12 +84,26 @@ export class AuthService {
 
   async refreshTokens(
     accessToken: string,
-    userInfo: UserType,
+    user: ITokenUser,
```
Comment (s-munro) [inline vb-be/src/modules/auth/auth.service.ts:88]: passing in `UserType` requires a bunch of unnecessary data here.  We only want the email, and perhaps the `id`, once we re-implement that.  Created a new interface to satisfy - lmk if you think it should be named differently

File: vb-be/src/util/error.ts [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
+  InvalidTokenPair = 'Invalid token pair',
   Generic = 'Something went wrong. Please try again later',
   TokenError = 'Cannot create tokens without user info',
   InvalidDataSet = 'Invalid dataset',
+  Unauthorized = 'Unauthorized',
 }
 
-export class ErrorHandler extends Error {
+// https://docs.nestjs.com/exception-filters
+export class ErrorHandler extends HttpException {
```
Comment (s-munro) [inline vb-be/src/util/error.ts:14]: nest uses a sort of innate error handler.  We can still use our system, we just had to change the class like this

File: vb-be/src/modules/auth/auth.service.ts (hunk 1 above)
Comment (hrishikeshpaul) [reply vb-be/src/modules/auth/auth.service.ts]: Yep it should be `vbUser`. I'm pushing a fix

Comment (hrishikeshpaul) [reply vb-be/src/modules/auth/auth.service.ts]: I think you can get the `id` back in the service for the token generation. It should be fine now

File: vb-be/src/modules/auth/auth.service.ts (hunk 2 above)
Comment (hrishikeshpaul) [reply vb-be/src/modules/auth/auth.service.ts:88]: yeah thats okay no problem with that

File: vb-be/src/util/error.ts (hunk 3 above)
Comment (hrishikeshpaul) [reply vb-be/src/util/error.ts:14]: ah cool

File: vb-be/src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -85,12 +84,26 @@ export class AuthService {
...
   ): Promise<string[]> {
-    if (!accessToken || !userInfo['email'] || !userInfo['id']) {
-      throw new ErrorHandler(HttpStatus.Error, ErrorText.InvalidRTArg);
-    }
     await this.delAsyncJwtClient(accessToken);
-    return await this.createTokens(userInfo);
+    return await this.createTokens(user);
+  }
+
+  async verifyToken(token: string, type: 'access' | 'refresh'): Promise<any> {
```
Comment (hrishikeshpaul) [inline vb-be/src/modules/auth/auth.service.ts]: use Enums here. keep it in the `auth.constants` like,
```
export enum TokenTypes {
Access = 'access',
Refresh = 'refresh,
}
```
use it like,
(token: string, type: TokenType)

File: vb-be/src/modules/auth/auth.service.ts [hunk 5]
```diff
@@ -85,12 +84,26 @@ export class AuthService {
...
+    let secret: string;
+    type === 'access'
```
Comment (hrishikeshpaul) [inline vb-be/src/modules/auth/auth.service.ts]: so this becomes type === TokenType.Access

**Comment by hrishikeshpaul** [comment]
So the `validate` route is like the `isLoggedIn` middleware correct? Because it checks whether the AT is valid. I think the `refresh` route is what i had in mind. I was thinking that the refresh route should just take in the AT/RT and the spotify AT/RT and refresh both the AT

**Comment by hrishikeshpaul** [comment]
Also i feel it'll be a good idea to make the exp of the refresh token infinite, i.e it doesn't expire. The reason we do this is cause we want to  keep the users logged in however long as they want. Even spotify does this

**Comment by s-munro** [comment]
Yeah /check is essentially isLoggedIn, but it's only checking the access token, since we're throwing a 403 to redirect to /refresh f the AT doesn't check out

**Comment by s-munro** [comment]
The refresh rote does take in both tokens, checks them, and returns new tokens. It doesn't have Spotify tokens set up yet because I wanted to make sure that was what we wanted after you woke up

**Comment by hrishikeshpaul** [comment]
> The refresh rote does take in both tokens, checks them, and returns new tokens. It doesn't have Spotify tokens set up yet because I wanted to make sure that was what we wanted after you woke up

ohhh dude my bad, i was just looking at the middleware didnt see the actual ROUTE. yeah ok imma try it out now

File: vb-be/src/modules/auth/auth.controller.ts [hunk 6]
```diff
@@ -84,6 +91,57 @@ export class AuthController {
...
+   * @return 403 (unauthorized) or 204 no content
+   */
+  @Get('/validate')
+  async validate(
+    @Headers('v-at') accessToken: string,
+    @Request() req: Req,
+    @Response() res: Res,
+  ) {
+    try {
+      const decoded = await this.authService.verifyToken(accessToken, 'access');
```
Comment (hrishikeshpaul) [inline vb-be/src/modules/auth/auth.controller.ts]: use the enum here aswell

**Comment by hrishikeshpaul** [comment]
looks good!

**Pull request merged**